src/RequestParser.cpp

```diff
@@ -0,0 +1,551 @@
+			index--;
+			break;
+		}
+		else if (!isValidURIChar(requestLine[index]) || requestLine[index] == '?') {
+			m_errorCode = 400;
+			throw std::runtime_error(ERR_URI_INVALID_CHAR);
+		}
+		else
+			m_request.uri.query.push_back(requestLine[index]);
+	}
+}
+
+/**
+ * @brief Parses the fragment component of the URI from the request line.
+ *
+ * This function extracts the fragment component of the URI from the provided request line.
+ * It starts parsing at the given index and continues until a space character is encountered.
+ * Each character in the fragment is validated using `isValidURIChar`.
+ * If an invalid character or another '#' is encountered,
+ * an error code is set, and a `std::runtime_error` is thrown.
+ *
+ * @param requestLine The request line string from which the fragment component of the URI is to be parsed.
+ * @param index The current index in the request line string where the fragment component starts.
+ *              This index is updated as the fragment is parsed.
+ * @throws std::runtime_error If an invalid character is encountered in the fragment,
+ *         or if another '#' is found, an error is thrown and the error code is set to 400.
+ */
+void	RequestParser::parseUriFragment(const std::string& requestLine, int& index)
+{
+	while (requestLine[++index]) {
+		if (requestLine[index] == ' ') {
+			index--;
+			break;
+		}
+		else if (!isValidURIChar(requestLine[index]) || requestLine[index] == '#') {
+			m_errorCode = 400;
+			throw std::runtime_error(ERR_URI_INVALID_CHAR);
+		}
+		else
+			m_request.uri.fragment.push_back(requestLine[index]);
+	}
+}
+
+/**
+ * @brief Parses the HTTP version from the request line.
+ *
+ * This function extracts and validates the HTTP version from the provided request line.
+ * It checks if the version format starts with "HTTP/",
+ * followed by a valid major and minor version number.
+ * If any format error is detected (e.g., missing digits or incorrect delimiter),
+ * an appropriate error code is set, and a `std::runtime_error` is thrown.
+ *
+ * @param requestLine The request line string from which the HTTP version is to be parsed.
+ * @return A substring of `requestLine` starting from the character after the parsed version.
+ * @throws std::runtime_error If the HTTP version format is invalid,
+ *         an error is thrown and the error code is set to 400.
+ */
+std::string	RequestParser::parseVersion(const std::string& requestLine)
+{
+	if (requestLine.substr(0, 5) != "HTTP/") {
+		m_errorCode = 400;
+		throw std::runtime_error(ERR_INVALID_VERSION_FORMAT);
+	}
+	int	i = 5;
+	if (!isdigit(requestLine[i])) {
+		m_errorCode = 400;
+		throw std::runtime_error(ERR_INVALID_VERSION_MAJOR);
+	}
+	m_request.version.push_back(requestLine[i]);
+	if (requestLine[++i] != '.') {
+		m_errorCode = 400;
+		throw std::runtime_error(ERR_INVALID_VERSION_DELIM);
+	}
+	m_request.version.push_back(requestLine[i]);
+	if (!isdigit(requestLine[++i])) {
```

Comment:
The same here as mentioned before, checking the value would be nice

Reply:
Agreed.